src/driver-sinowealth.c

```diff
@@ -888,6 +1552,49 @@ sinowealth_write_config(struct ratbag_device *device)
 	return 0;
 }
 
+static int
+sinowealth_write_macros(struct ratbag_device *device)
+{
+	int rc = 0;
+	struct ratbag_button *button = NULL;
+	struct ratbag_profile *profile = NULL;
+
+	struct sinowealth_data *drv_data = device->drv_data;
+
+	const char config_report_id = drv_data->is_long ? SINOWEALTH_REPORT_ID_CONFIG_LONG : SINOWEALTH_REPORT_ID_CONFIG;
+
+	/* NOTE: We will reuse the same buffer for all commands and reset it's `events` field for every button. */
+	struct sinowealth_macro_report macro = {};
+	macro.report_id = config_report_id;
+	macro.command_id = SINOWEALTH_CMD_MACRO;
+	macro.unknown1 = 0x2;
+
+	ratbag_device_for_each_profile(device, profile) {
+		ratbag_profile_for_each_button(profile, button) {
+			if (!button->dirty)
+				continue;
+
+			struct ratbag_button_action *action = &button->action;
+
+			/* Ignore non macro actions.
+			 * They were already handled by @ref sinowealth_update_profile_from_buttons.
+			 */
+			if (action->type != RATBAG_BUTTON_ACTION_TYPE_MACRO)
+				continue;
+
+			sinowealth_update_macro_from_action(profile, button, &macro);
+
+			rc = sinowealth_query_write(device, (uint8_t*)&macro, sizeof(macro));
+			if (rc != 0) {
+				log_error(device->ratbag, "Error while writing macro %u: %d\n", macro.index, rc);
+				return -1;
```

Comment:
```suggestion
				return -EINVAL;
```

Reply:
Same as above - To be honest I don't think EINVAL fits here as it can be any kind of error.